core/client/hammer/hammer.go

```diff
@@ -76,125 +97,199 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
+		executeRequests(ctx, args, handlers)
+		fmt.Print("\n")
+	}
 
-	return outputs
+	return nil //pool.Stop()
 }
 
-// gen produces a channel of integers from 0 up to count-1 and then closes.
-func gen(ctx context.Context, count int) <-chan int {
-	jobs := make(chan int)
+func genArgs(ctx context.Context, qps, batch, count int, duration time.Duration) <-chan opArg {
+	inflightReqs := make(chan opArg, qps)
 	go func() {
-		defer close(jobs)
+		cctx, cancel := context.WithTimeout(ctx, duration)
+		defer cancel()
+		defer close(inflightReqs)
+
+		rateLimiter := rate.NewLimiter(rate.Limit(qps), qps)
 		for i := 0; i < count; i++ {
-			select {
-			case <-ctx.Done():
+			userIDs := make([]string, 0, batch)
+			for j := 0; j < batch; j++ {
+				userIDs = append(userIDs, fmt.Sprintf("user_%v", i*batch+j))
+			}
+			if err := rateLimiter.WaitN(cctx, batch); err != nil {
+				log.Printf("Duration %v elapsed, stopping request generation", duration)
 				return
-			case jobs <- i:
+			}
+			inflightReqs <- opArg{
+				UserIDs:  userIDs,
+				PageSize: batch,
 			}
 		}
 	}()
-	return jobs
+	return inflightReqs
 }
 
-// printStats prints stats repeatedly every period until outputs closes.
-func (h *Hammer) printStats(period time.Duration, outputs <-chan error) {
-	startTime := time.Now()
-	var totalRequests float64
-
-	// Print stats asyncronously until refresh closes.
-	refresh := time.NewTicker(period)
-	stop := make(chan struct{})
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for {
-			select {
-			case <-refresh.C:
-				rate := totalRequests / time.Since(startTime).Seconds()
-				log.Printf("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
-			case <-stop:
-				rate := totalRequests / time.Since(startTime).Seconds()
-				log.Printf("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
-				refresh.Stop()
-				return
-			}
-		}
-	}()
+type worker struct {
+	*Hammer
+	client *client.Client
+	bar    *cpb.ProgressBar
+}
 
-	// Collect outputs until output closes.
-	for err := range outputs {
-		totalRequests++
+func (h *Hammer) newWorkers(n int) ([]worker, error) {
+	workers := make([]worker, 0, n)
+	for i := 0; i < n; i++ {
+		// Give each worker its own client.
+		client, err := client.NewFromConfig(h.ktCli, h.domain)
 		if err != nil {
-			log.Printf("writeOp(): %v", err)
+			return nil, err
 		}
+
+		workers = append(workers, worker{
+			Hammer: h,
+			client: client,
+			bar:    cpb.New(100).Prefix(fmt.Sprintf("%v ", i)),
+		})
 	}
-	close(stop)
-	wg.Wait()
+	return workers, nil
 }
 
-func (h *Hammer) worker(ctx context.Context, jobs <-chan int, outputs chan<- error) {
-	client, err := client.NewFromConfig(h.ktCli, h.config)
-	if err != nil {
-		glog.Errorf("NewFromConfig(): %v", err)
-		return
+// writeOp queues many user mutations, waits, and then verifies them all.
+func (w *worker) writeOp(ctx context.Context, req *opArg) error {
+	users := make([]*tpb.User, 0, len(req.UserIDs))
```

Comment:
It appears that this `users` array is created only to be scanned below. Maybe merge the two loops?

```
for _, userID := range req.UserIDs {
  callOptions := w.callOptions(userID)
  u := &tpb.User{
    DomainId:       w.domain.DomainId,
    AppId:          w.appID,
    UserId:         userID,
    PublicKeyData:  []byte("publickey"),
    AuthorizedKeys: w.authorizedKeys,
  }
  ...
}
```

Reply:
Splitting the loops allows the changes to be efficiently batched. 
If there was only one loop, we would be waiting for a new map revision with one one change in it.